lib/Dialect/AIE/IR/AIEDialect.cpp

```diff
@@ -1545,60 +1545,107 @@ LogicalResult DMABDOp::verify() {
+
+  if (getLenInBytes() % 4)
+    return emitOpError("transfer length must be multiple of 4B (32b)");
+
+  if (getOperation()->getParentOfType<MemOp>()) {
+    auto memOp = getOperation()->getParentOfType<MemOp>();
     if (auto bufferOp = getBufferOp();
         bufferOp.getTileOp().colIndex() != memOp.colIndex() ||
         bufferOp.getTileOp().rowIndex() != memOp.rowIndex())
-      return emitOpError("can only access a buffer in the same tile.");
+      return emitOpError(
+          "Core tile DMAs can only access a buffer in the same tile.");
+    if (std::optional<int32_t> bdId = getBdId(); bdId.has_value() && *bdId > 15)
+      return emitOpError("Core tile DMAs have at most 16 buffer descriptors");
+    if (std::optional<int32_t> nextBdId = getNextBdId();
+        nextBdId.has_value() && *nextBdId > 15)
+      return emitOpError("Core tile DMAs have at most 16 buffer descriptors");
+  } else if (getOperation()->getParentOfType<ShimDMAOp>()) {
+    if (!isa<ExternalBufferOp>(getBuffer().getDefiningOp()))
+      return emitOpError("Shim tile DMAs can only access external buffers");
+    if (std::optional<int32_t> bdId = getBdId(); bdId.has_value() && *bdId > 15)
+      return emitOpError("Shim tile DMAs have at most 16 buffer descriptors");
+    if (std::optional<int32_t> nextBdId = getNextBdId();
+        nextBdId.has_value() && *nextBdId > 15)
+      return emitOpError("Shim tile DMAs have at most 16 buffer descriptors");
+  } else {
+    assert(getOperation()->getParentOfType<MemTileDMAOp>() &&
+           "expected MemTileDMAOp parent");
+    auto memOp = getOperation()->getParentOfType<MemTileDMAOp>();
+    if (auto bufferOp = getBufferOp();
+        bufferOp.getTileOp().rowIndex() != memOp.rowIndex() ||
+        std::abs(bufferOp.getTileOp().colIndex() - memOp.colIndex()) > 1)
+      return emitOpError("Mem tile DMAs can only access adjacent mem tiles");
+    // TODO(max): check even channel -> <24; odd channel -> > 24
+    if (std::optional<int32_t> bdId = getBdId(); bdId.has_value() && *bdId > 47)
+      return emitOpError("Memtile DMAs have at most 48 buffer descriptors");
+    if (std::optional<int32_t> nextBdId = getNextBdId();
+        nextBdId.has_value() && *nextBdId > 47)
+      return emitOpError("Memtile DMAs have at most 48 buffer descriptors");
   }
 
   // The following checks only apply if non-default strides/wraps are defined.
   if (getDimensions()) {
-    MemRefType buffer = getBuffer().getType();
-    // We are not restrictive about the type of the memref used as the input
-    // to the DMABD when used with multi-dimensional strides/wraps. Since the
-    // BD will use the memref as a base address and copy from it in 32 bit
-    // chunks, while assuming the layout of the memref is contiguous. We
-    // assume the user/compiler understands and accounts for this.
-    uint64_t memrefSize = 1; // in bytes
-    uint64_t maxIdx = 0;
-    for (int64_t memrefDim : buffer.getShape())
-      memrefSize *= 4 * memrefDim;
```

Comment:
Dunno what this is about but it's wrong for at least 2 reasons (assumes 4 bytes per element, and then for some reason computes a "byte volume"???).

Reply:
That seems odd... it seems like there's too many '4's getting multiplied?

Reply:
n-d memory! `bytes^n`!